### PR TITLE
Fixes #587 Support negative numbers as command option values

### LINKED ISSUE #587: Cannot specify negative number as command option value
As of version 0.40.0, it doesn't appear possible to specify a negative number as a command option.

Assuming I define a command that takes a decimal value as a command option, and that I want to pass a negative number value, neither of the following works:
```
mycommand -a -1.50
mycommand -a "-1.50"
```

Both attempts fail with this error `Not a valid name for a short option`. It appears that the command option parsing logic thinks that the -1.50 is supposed to be a command option.

I would expect that the quotes at least would effectively escape the numeric value. Is this a bug, or am I missing something?

Thanks.

### PR DESCRIPTION
**Pull request by jouniheikniemi**
Decided to create some unit tests for the tokenizer, as it would've been too much of a mess to test all of this through CommandApp. That's obviously up for debate, but at least the tokenizer tests helped me with this.

### CODE REVIEW

**Review by phil-scott-78 — COMMENTED** [review]
Had some time to look through this. At a glance it looks good. But think we could get some tests that cover the invalid cases as described in the comments of the tokenizer for invalid items?

**Comment by jouniheikniemi** [comment]
Hey @phil-scott-78, thanks for reviewing. Actually, CommandTreeTokenizer does have (logical) 100 % test coverage already. It is not immediately obvious, as this PR is the first stage that actually adds tests specifically for the Tokenizer; previously, the tokenizer has only been tested through CommandAppTests which test the entire parsing and execution pipeline. However, some of the test scenarios there (say, CommandAppTests.Parsing.InvalidShortOptionName.Should_Return_Correct_Text) do exercise these failure modes.

I would personally prefer to have 100 % coverage specifically with the tokenizer test suite, but writing all that is slightly beyond the scope of this issue. Would you expect to see (tokenizer level) tests for the failure scenarios of the ScanShortOptions tokenizer branch or what? TBH, the tokenizer test set in this PR doesn't cover all short option tokenization features (even the happy paths) either, grouped properties probably being the most important missing part. Given that there's no precedent for tokenizer tests in the project, I'm happy to do extra work to set a good example, let's just agree on the goal first.

File: src/Spectre.Console/Cli/Internal/Parsing/CommandTreeTokenizer.cs [hunk 1]
```diff
@@ -235,13 +235,34 @@ private static IEnumerable<CommandTreeToken> ScanShortOptions(CommandTreeTokeniz
...
+                result.Add(new CommandTreeToken(CommandTreeToken.Kind.String, position, value, value));
+            }
             else
             {
-                // Create a token representing the short option.
-                var tokenPosition = position + 1 + result.Count;
-                var represntation = current.ToString(CultureInfo.InvariantCulture);
-                var token = new CommandTreeToken(CommandTreeToken.Kind.ShortOption, tokenPosition, represntation, represntation);
-                throw CommandParseException.InvalidShortOptionName(reader.Original, token);
+                ThrowInvalidShortOptionName(current.ToString(CultureInfo.InvariantCulture));
```
Comment (FrankRay78) [inline src/Spectre.Console/Cli/Internal/Parsing/CommandTreeTokenizer.cs:265]: I do not see the benefit of putting the local three-line logic into a separate private method, which also still references `reader.Original` from the calling function without passing this in as an explicit argument.  I suggest this private method be removed and the logic relocated inline here, instead.

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+namespace Spectre.Console.Tests.Unit.Cli;
+
+public static class CommandTreeTokenizerTests
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:3]: Fantastic that you have introduced explicit unit tests for the tokenizer! A positive move away from the general approach of 'unit tests' being located at the CommandApp level (nb. more like integration tests really).

Does this class need to be static?  Unit test classes aren't generally made static - suggest this is also removed to ensure consistency with the rest of the codebase.

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            t.Position.ShouldBe(0);
+            t.Value.ShouldBe("a");
+            t.Representation.ShouldBe("-a");
+        }
+
+        [Theory]
+        [InlineData("-a:foo")]
+        [InlineData("-a=foo")]
+        public void With_Value(string param)
+        {
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:25]: // When

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            var result = CommandTreeTokenizer.Tokenize(new[] { param });
+            Assert.Empty(result.Remaining);
+            Assert.Equal(2, result.Tokens.Count);
+
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:29]: // Then

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            t.Value.ShouldBe("foo");
+            t.Representation.ShouldBe("foo");
+        }
+
+        [Theory]
+        [InlineData("-a:-1.5", null)]
+        [InlineData("-a=-1.5", null)]
+        [InlineData("-a", "-1.5")]
+        public void With_Negative_Numeric_Value(string firstArg, string secondArg)
+        {
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:50]: // Given

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            List<string> args = new List<string>();
+            args.Add(firstArg);
+            if (secondArg != null)
+            {
+                args.Add(secondArg);
+            }
+
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:57]: // When

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            var result = CommandTreeTokenizer.Tokenize(args);
+            Assert.Empty(result.Remaining);
+            Assert.Equal(2, result.Tokens.Count);
+
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:61]: // Then

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            t.TokenKind.ShouldBe(CommandTreeToken.Kind.String);
+            t.IsGrouped.ShouldBe(false);
+            t.Position.ShouldBe(3);
+            t.Value.ShouldBe("-1.5");
+            t.Representation.ShouldBe("-1.5");
+        }
+
+        [Fact]
+        public void With_Negative_Numeric_Prefixed_String_Value()
+        {
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:79]: // When

File: test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            var result = CommandTreeTokenizer.Tokenize(new[] { "-6..2 " });
+            Assert.Empty(result.Remaining);
```
Comment (FrankRay78) [inline test/Spectre.Console.Tests/Unit/Cli/CommandTreeTokenizerTests.cs:81]: // Then

**Review by FrankRay78 — CHANGES_REQUESTED** [review]
Thank you for submitting this PR, my review comments are largely stylistic to ensure the changes fit stylistically with the rest of the codebase.

**Comment by FrankRay78** [comment]
> this PR is the first stage that actually adds tests specifically for the Tokenizer; previously, the tokenizer has only been tested through CommandAppTests which test the entire parsing and execution pipeline

This is a significant improvement that @jouniheikniemi has introduced and I would be keen to see this PR, once review comments are addressed 😊, merged into the codebase.  Particularly since I have a raft of CommandTreeTokenizer tests I would like to add myself, ideally into the same class being introduced in this PR.

**Comment by FrankRay78** [comment]
This should be closed as it's been fully included in https://github.com/spectreconsole/spectre.console/pull/1048  [I don't have permission to close another person's PR]

**Comment by jouniheikniemi** [comment]
I was about to get back to your review comments, but it seems you'd rather pick up the ball yourself; fine with me. Closing the PR. Thanks!

**Comment by FrankRay78** [comment]
I was keen to merge your excellent work @jouniheikniemi - thanks for your contribution.

**Pull request closed without merging**